client/my-sites/plugins/marketplace/marketplace-domain-upsell/index.tsx

```diff
@@ -0,0 +1,204 @@
+				icon
+			>
+				{ translate( 'Domain names and their influence on SEO' ) }
+			</ExternalLink>
+			<HorizontalRule />
+		</div>
+	);
+}
+
+const getSiteNameFromURL = ( url: string ) => url?.split( '.' )?.[ 0 ] ?? url;
+
+function CalypsoWrappedMarketplaceDomainUpsell(): JSX.Element {
+	const [ selectedDomainProductUUID, setDomainProductUUID ] = useState< string >( '' );
+	const [ selectedDomain, setDomain ] = useState< DomainSuggestions.DomainSuggestion | null >(
+		null
+	);
+
+	const [ isExpandedBasketView, setIsExpandedBasketView ] = useState( false );
+	const { addProductsToCart, replaceProductsInCart, removeProductFromCart } = useShoppingCart();
+	const products = useSelector( getProductsList );
+	const previousPath = useSelector( getPreviousPath );
+	const isFetchingProducts = useSelector( isProductsListFetching );
+	const selectedSite = useSelector( getSelectedSite );
+	const { domain } = useSelector( getWpComDomainOfSelectedSite );
+	const siteName = getSiteNameFromURL( domain );
+	const dispatch = useDispatch();
+
+	useEffect( () => {
+		setIsExpandedBasketView( isDesktop() );
+		selectedSite && dispatch( fetchSiteDomains( selectedSite.ID ) );
+	}, [ setIsExpandedBasketView, selectedSite, dispatch ] );
+
+	useEffect( () => {
+		//FIXME: This code segment simulates yoast premium already being added when arriving here. To be removed when plugins page is completed.
+		if ( ! isFetchingProducts && products[ 'yoast_premium' ] ) {
+			const yoastProduct = fillInSingleCartItemAttributes(
+				{ product_slug: 'yoast_premium' },
+				products
+			);
+			replaceProductsInCart( [ yoastProduct ] );
+		}
+	}, [ isFetchingProducts, products, replaceProductsInCart ] );
+
+	const redirectToUseDomainFlow = (): void => {
+		const currentUrl = '/plugins/domain';
+		const useYourDomainUrl = `/start/new-launch/domains-launch/use-your-domain?siteSlug=${ selectedSite?.slug }&source=${ currentUrl }`;
+		page( useYourDomainUrl );
+	};
+
+	const onDomainSelect = ( suggestion: DomainSuggestions.DomainSuggestion ) => {
+		const { product_slug, domain_name } = suggestion;
+		const domainProduct = {
+			...domainRegistration( {
+				productSlug: product_slug,
+				domain: domain_name,
+				source: 'Marketplace-Yoast-Domain-Upsell',
+			} ),
+			...suggestion,
+		};
+		setDomain( domainProduct );
+
+		//First remove the previously added domain
+		new Promise< void >( ( resolve ) => {
+			if ( ! selectedDomainProductUUID ) {
+				resolve();
+			} else {
+				removeProductFromCart( selectedDomainProductUUID ).then( () => {
+					setDomainProductUUID( '' );
+					resolve();
+				} );
+			}
+		} ).then( () => {
+			//Then add the new domain
+			addProductsToCart( [ domainProduct ] ).then( ( responseCart: ResponseCart ) => {
+				const productAdded = responseCart.products.find(
+					( { product_slug: added_product_slug } ) => added_product_slug === product_slug
+				);
+				setDomainProductUUID( productAdded?.uuid ?? '' );
+			} );
+		} );
```

Comment:
This code would benefit greatly from using `async`/`await`:
```js
if ( selectedDomainProductUUID ) {
  await removeProductFromCart();
  setDomainProductUUID( '' );
}

const responseCart = await addProductsToCart();
const productAdded = responseCart.products.find( ... );
```

Reply:
Oh cool I have not used async await conditionally before! Thanks for this handy tip, this looks much cleaner!